src/providers/azure.rs

```diff
@@ -0,0 +1,503 @@
+            publicIpAddress: String,
+            resourceGroup: String,
+        }
+
+        let out = Command::new("az")
+            .args(&[
+                "vm",
+                "create",
+                "--resource-group",
+                rg,
+                "--name",
+                name,
+                "--image",
+                "UbuntuLTS",
+                "--size",
+                size,
+                "--admin-username",
+                username,
+                "--ssh-key-value",
+                "@~/.ssh/id_rsa.pub",
+            ])
+            .output()?;
+
+        if !out.status.success() {
+            return Err(format_err!("Failed to create vm {}", name))
+                .context(String::from_utf8(out.stderr).unwrap())?;
+        }
+
+        let vm: VmCreateOut = serde_json::from_slice(&out.stdout)?;
+        ensure!(vm.powerState == "VM running", "VM power state incorrect");
+        ensure!(vm.resourceGroup == rg, "VM resource group incorrect");
+        Ok(vm.publicIpAddress)
+    }
+
+    pub fn open_ports(rg: &str, vm_name: &str) -> Result<(), Error> {
+        let out = Command::new("az")
+            .args(&[
+                "vm",
+                "open-port",
+                "--port",
+                "0-65535",
+                "--resource-group",
+                rg,
+                "--name",
+                vm_name,
+            ])
+            .output()?;
+        if !out.status.success() {
+            return Err(format_err!("Failed to open ports for {}", vm_name))
+                .context(String::from_utf8(out.stderr).unwrap())?;
+        }
+
+        Ok(())
+    }
+
+    pub fn delete_resource_group(rg: &str) -> Result<(), Error> {
+        let out = Command::new("az")
+            .args(&["group", "delete", "--name", rg, "--yes"])
+            .output()?;
+        if !out.status.success() {
+            bail!("Failed to delete resource group {}", rg)
+        }
+
+        Ok(())
+    }
+}
+
+/// A descriptor for a single Azure VM type. Only UbuntuLTS VMs are supported.
+#[derive(Clone)]
+pub struct Setup {
+    region: Region,
+    instance_type: String,
+    setup_fn:
+        Option<Arc<dyn Fn(&mut ssh::Session, &slog::Logger) -> Result<(), Error> + Send + Sync>>,
+}
+
+impl Default for Setup {
+    fn default() -> Self {
+        Setup {
+            region: "eastus".parse().unwrap(),
+            instance_type: "Standard_DS1_v2".to_string(),
+            setup_fn: None,
+        }
+    }
+}
+
+impl super::MachineSetup for Setup {
+    type Region = Region;
+
+    fn region(&self) -> Self::Region {
+        self.region
+    }
+}
+
+impl Setup {
+    pub fn region(self, r: Region) -> Self {
+        Self { region: r, ..self }
+    }
+
+    /// To view the available sizes in the relevant region, use:
+    /// `az vm list-sizes -l <region_name>`.
+    pub fn instance_type(self, inst_type: String) -> Result<Self, Error> {
+        if azcmd::available_instances_in_region(self.region)?
```

Comment:
I don't think the AWS version checks that the instance type is valid this early? Any particular reason you want to do it here rather than in `launch` like AWS does?

Reply:
This is probably a better place to put the check overall - it's more logical that setting a descriptor field causes an error, rather than allowing a bad descriptor field and then failing later.
Azure has an easy api that lists the available types, so why not call it here?

Reply:
If so, I think we should add the check in the same place for AWS :)
The reason not to do it is that I don't think people expect builder methods to block on a network connection.

Reply:
yeah, the network request is somewhat unfortunate (although I think `az` does some caching).

`az vm create` will fail if you give it a bad instance_type anyway, so might as well delete some code.